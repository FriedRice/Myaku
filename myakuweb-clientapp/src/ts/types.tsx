/* Types used across the MyakuWeb project */

export type PrimativeType = string | number | symbol | null | undefined;

export function isPrimativeType(value: unknown): value is PrimativeType {
    return (
        value === null
        || (typeof value !== 'object'
            && typeof value !== 'function')
    );
}

export interface Indexable {
    [key: string]: unknown;
}

export function isIndexable(value: unknown): value is Indexable {
    return !isPrimativeType(value);
}

export type KanaConvertType = 'hira' | 'kata' | 'none';

export const KANA_CONVERT_TYPE_VALUES: KanaConvertType[] = [
    'hira',
    'kata',
    'none',
];

export function isKanaConvertType(value: unknown): value is KanaConvertType {
    if (typeof value !== 'string') {
        return false;
    }
    return KANA_CONVERT_TYPE_VALUES.includes(value as KanaConvertType);
}

export interface SearchOptions {
    kanaConvertType: KanaConvertType;
}

export const SEARCH_OPTIONS: Array<keyof SearchOptions> = [
    'kanaConvertType',
];

export const DEFAULT_SEARCH_OPTIONS: SearchOptions = {
    kanaConvertType: 'hira',
};

export function isSearchOption(value: unknown): value is keyof SearchOptions {
    if (typeof value !== 'string') {
        return false;
    }
    return SEARCH_OPTIONS.includes(value as keyof SearchOptions);
}

export const enum PageDirection {
    Next = 'next',
    Previous = 'previous',
}

export interface Search {
    query: string;
    pageNum: number;
    options: SearchOptions;
}

export interface SessionSearchOptionsResponse {
    readonly kanaConvertType: KanaConvertType;
}

export interface ArticleSampleTextSegment {
    isQueryMatch: boolean;
    text: string;
}

export interface ArticleSampleText {
    textStartPos: string;
    segments: ArticleSampleTextSegment[];
}

export interface ArticleSearchResult {
    articleId: string;
    title: string;
    sourceName: string;
    sourceUrl: string;
    publicationDatetime: Date;
    lastUpdatedDatetime: Date | null;
    instanceCount: number;
    tags: string[];
    mainSampleText: ArticleSampleText;
    moreSampleTexts: ArticleSampleText[];
}

export interface SearchResultPage {
    search: Search;
    totalResults: number;
    hasNextPage: boolean;
    maxPageReached: boolean;
    results: ArticleSearchResult[];
}

export interface SearchResultPageResponse {
    readonly convertedQuery: string;
    readonly totalResults: number;
    readonly pageNum: number;
    readonly hasNextPage: boolean;
    readonly maxPageReached: boolean;
    readonly articleResults: ArticleSearchResult[];
}

export interface ResourceLink {
    resourceName: string;
    link: string;
}

export interface ResourceLinkSet {
    setName: string;
    resourceLinks: ResourceLink[];
}

export interface SearchResources {
    query: string;
    resourceLinkSets: ResourceLinkSet[];
}

export interface ResourceLinksResponse {
    readonly convertedQuery: string;
    readonly resourceLinkSets: ResourceLinkSet[];
}
