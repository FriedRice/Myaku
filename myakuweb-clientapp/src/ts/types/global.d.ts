/**
 * Global types for loading webpack non-code assets into TypeScript files.
 */

/**
 * Svg image files import as the URL to that image.
 */
declare module '*.svg' {
    const sourceUrl: string;
    export default sourceUrl;
}
