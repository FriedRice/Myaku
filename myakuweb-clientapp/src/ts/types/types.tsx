/**
 * Types used across the MyakuWeb app.
 */

/** Makes the type of every key of an object nullable. */
export type AllNullable<T> = {
    [P in keyof T]: T[P] | null;
};

/** The built-in non-object types in javascript. */
export type PrimativeType = (
    string | boolean | number | symbol | null | undefined
);

/**
 * Type guard function for checking that a value is a [[PrimativeType]].
 */
export function isPrimativeType(value: unknown): value is PrimativeType {
    return (
        value === null
        || (typeof value !== 'object'
            && typeof value !== 'function')
    );
}

/** Type that supports indexing to get some value (e.g. objects). */
export interface Indexable {
    [key: string]: unknown;
}

/**
 * Type guard function for checking that a value is a [[Indexable]].
 */
export function isIndexable(value: unknown): value is Indexable {
    return !isPrimativeType(value);
}

/** Possible conversion types between romaji and kana. */
export type KanaConvertType = 'hira' | 'kata' | 'none';

/** Array containing all of the valid [[KanaConvertType]] values. */
export const KANA_CONVERT_TYPE_VALUES: KanaConvertType[] = [
    'hira',
    'kata',
    'none',
];

/**
 * Type guard function for checking that a value is a [[KanaConvertType]].
 */
export function isKanaConvertType(value: unknown): value is KanaConvertType {
    if (typeof value !== 'string') {
        return false;
    }
    return KANA_CONVERT_TYPE_VALUES.includes(value as KanaConvertType);
}

/**
 * Search options that can be optionally specified along with a [[Search]].
 */
export interface SearchOptions {
    kanaConvertType: KanaConvertType;
}

/** Array containing all of the valid [[SearchOptions]] keys. */
export const SEARCH_OPTIONS: Array<keyof SearchOptions> = [
    'kanaConvertType',
];

/**
 * Type guard function for checking that a value is a [[SearchOptions]] key.
 */
export function isSearchOption(value: unknown): value is keyof SearchOptions {
    if (typeof value !== 'string') {
        return false;
    }
    return SEARCH_OPTIONS.includes(value as keyof SearchOptions);
}

/**
 * Type assert function for asserting that a value is a [[SearchOptions]] key.
 */
export function assertIsSearchOption(
    value: unknown
): asserts value is keyof SearchOptions {
    if (!isSearchOption(value)) {
        throw new Error(`"${value}" is not a valid search option`);
    }
}

/** The required parameters to make a MyakuWeb search */
export interface Search {
    query: string;
    pageNum: number;
}

/** A single segment of text from an [[ArticleSampleText]]. */
export interface ArticleSampleTextSegment {
    isQueryMatch: boolean;
    text: string;
}

/** Sample text for an article for an [[ArticleSearchResult]]. */
export interface ArticleSampleText {
    textStartPos: string;
    segments: ArticleSampleTextSegment[];
}

/** A single search result for a MyakuWeb search. */
export interface ArticleSearchResult {
    articleId: string;
    title: string;
    sourceName: string;
    sourceUrl: string;
    publicationDatetime: Date;
    lastUpdatedDatetime: Date | null;
    instanceCount: number;
    tags: string[];
    mainSampleText: ArticleSampleText;
    moreSampleTexts: ArticleSampleText[];
}

/** A page of search results for a MyakuWeb search. */
export interface SearchResultPage {
    search: Search;
    totalResults: number;
    hasNextPage: boolean;
    maxPageReached: boolean;
    results: ArticleSearchResult[];
}

/**
 * The search results page response JSON format for the MyakuWeb search API.
 */
export interface SearchResultPageResponse {
    readonly totalResults: number;
    readonly pageNum: number;
    readonly hasNextPage: boolean;
    readonly maxPageReached: boolean;
    readonly articleResults: ArticleSearchResult[];
}

/** A single resource link from a [[ResourceLinkSet]]. */
export interface ResourceLink {
    resourceName: string;
    link: string;
}

/** A set of resource links to accompany a MyakuWeb search. */
export interface ResourceLinkSet {
    setName: string;
    resourceLinks: ResourceLink[];
}

/** Resources to accompany a MyakuWeb search. */
export interface SearchResources {
    query: string;
    resourceLinkSets: ResourceLinkSet[];
}

/** The search resources response JSON format for the MyakuWeb search API. */
export interface ResourceLinksResponse {
    readonly resourceLinkSets: ResourceLinkSet[];
}
