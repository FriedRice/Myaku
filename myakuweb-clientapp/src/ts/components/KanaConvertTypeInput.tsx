/** @module Romaji->kana conversion setting input component */

import React from 'react';

import {
    KANA_CONVERT_TYPE_VALUES,
    KanaConvertType,
} from '../types';

interface KanaConvertTypeInputProps {
    kanaConvertType: KanaConvertType;
    onChange: (kanaConvertType: KanaConvertType) => void;
}
type Props = KanaConvertTypeInputProps;

const CONVERT_TYPE_LABELS: Record<KanaConvertType, string> = {
    'hira': 'Hiragana (a→あ)',
    'kata': 'Katakana (a→ア)',
    'none': 'No conversion (a→a)',
};


function createRadioInputs(
    selected: KanaConvertType,
    handleChangeFunc: (event: React.FormEvent<HTMLInputElement>) => void
): React.ReactElement[] {
    var radioInputs: React.ReactElement[] = [];
    for (const convertType of KANA_CONVERT_TYPE_VALUES) {
        const inputId = `kana-conv-${convertType}-radio`;
        radioInputs.push(
            <div key={inputId} className='search-options-check'>
                <input
                    className='search-options-check-input'
                    id={inputId}
                    type='radio'
                    name='kana-convert-type'
                    value={convertType}
                    checked={convertType === selected}
                    onChange={handleChangeFunc}
                />
                <label
                    className='check-input-label'
                    htmlFor={inputId}
                >
                    {CONVERT_TYPE_LABELS[convertType]}
                </label>
            </div>
        );
    }

    return radioInputs;
}

const KanaConvertTypeInput: React.FC<Props> = function(props) {
    function handleChange(event: React.FormEvent<HTMLInputElement>): void {
        props.onChange(event.currentTarget.value as KanaConvertType);
    }

    return (
        <fieldset className='kana-conv-field'>
            <legend className='search-options-legend'>
                Romaji Conversion
            </legend>
            {createRadioInputs(props.kanaConvertType, handleChange)}
        </fieldset>
    );
};

export default KanaConvertTypeInput;
