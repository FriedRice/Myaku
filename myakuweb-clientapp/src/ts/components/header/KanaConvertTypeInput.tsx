/** @module Romaji->kana conversion setting input component */

import React from 'react';
import useInputChangeHandler from 'ts/hooks/useInputChangeHandler';

import {
    KANA_CONVERT_TYPE_VALUES,
    KanaConvertType,
} from 'ts/types/types';

interface KanaConvertTypeInputProps {
    kanaConvertType: KanaConvertType;
    onChange: (kanaConvertType: KanaConvertType) => void;
}
type Props = KanaConvertTypeInputProps;

const CONVERT_TYPE_LABELS: Record<KanaConvertType, string> = {
    'hira': 'Hiragana (a→あ)',
    'kata': 'Katakana (a→ア)',
    'none': 'No conversion (a→a)',
};


function createRadioInputs(
    selected: KanaConvertType,
    handleChangeFunc: (event: React.FormEvent<HTMLInputElement>) => void
): React.ReactElement[] {
    var radioInputs: React.ReactElement[] = [];
    for (const convertType of KANA_CONVERT_TYPE_VALUES) {
        const inputId = `kana-conv-${convertType}-radio`;
        radioInputs.push(
            <div key={inputId} className='search-options-check'>
                <input
                    className='search-options-check-input'
                    id={inputId}
                    type='radio'
                    name='kana-convert-type'
                    value={convertType}
                    checked={convertType === selected}
                    onChange={handleChangeFunc}
                />
                <label
                    className='check-input-label'
                    htmlFor={inputId}
                >
                    {CONVERT_TYPE_LABELS[convertType]}
                </label>
            </div>
        );
    }

    return radioInputs;
}

const KanaConvertTypeInput: React.FC<Props> = function(props) {
    const handleChange = useInputChangeHandler(props.onChange);

    return (
        <fieldset className='kana-conv-field'>
            <legend className='search-options-legend'>
                Romaji Conversion
            </legend>
            {createRadioInputs(props.kanaConvertType, handleChange)}
        </fieldset>
    );
};

export default KanaConvertTypeInput;
