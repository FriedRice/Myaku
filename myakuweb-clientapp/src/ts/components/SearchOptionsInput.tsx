/** @module Wrapper for all search options input components  */

import KanaConvertTypeInput from './KanaConvertTypeInput';
import React from 'react';
import { KanaConvertType, SearchOptions } from '../types';

interface OnChangeFunc {
    <K extends keyof SearchOptions> /* </> */ (
        changedOption: K, newValue: SearchOptions[K]
    ): void;
}

interface SearchOptionsInputProps {
    searchOptions: SearchOptions;
    onChange: OnChangeFunc;
}
type Props = SearchOptionsInputProps;


const SearchOptionsInput: React.FC<Props> = function(props) {
    function handleKanaConvertTypeChange(newValue: KanaConvertType): void {
        props.onChange<'kanaConvertType'>('kanaConvertType', newValue);
    }

    return (
        <KanaConvertTypeInput
            kanaConvertType={props.searchOptions.kanaConvertType}
            onChange={handleKanaConvertTypeChange}
        />
    );
};

export default SearchOptionsInput;
