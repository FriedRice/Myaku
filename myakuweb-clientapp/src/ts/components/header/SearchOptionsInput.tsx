/**
 * SearchOptionsInput component module. See [[SearchOptionsInput]].
 */

import KanaConvertTypeInput from 'ts/components/header/KanaConvertTypeInput';
import React from 'react';

import { KanaConvertType, SearchOptions } from 'ts/types/types';

/**
 * Generic search option onChange function that takes any valid key-value
 * search option pair as its arguments.
 */
interface OnChangeFunc {
    // </> comment is to stop JSX syntax highlighting from applying
    <K extends keyof SearchOptions> /* </> */ (
        changedOption: K, newValue: SearchOptions[K]
    ): void;
}

/** Props for the [[SearchOptionsInput]] component. */
interface SearchOptionsInputProps {
    /**
     * Current inputted search options for the form containing this component.
     *
     * These options will be displayed as selected in the search option inputs
     * rendered by the component.
     */
    searchOptions: SearchOptions;

    /**
     * Callback to call with a new value for a search option whenever any
     * search option value changes.
     */
    onChange: OnChangeFunc;
}
type Props = SearchOptionsInputProps;


/**
 * Wrapper component for all search option input components.
 *
 * @param props - See [[SearchOptionsInputProps]].
 */
const SearchOptionsInput: React.FC<Props> = function(props) {
    function handleKanaConvertTypeChange(newValue: KanaConvertType): void {
        props.onChange<'kanaConvertType'>('kanaConvertType', newValue);
    }

    return (
        <KanaConvertTypeInput
            kanaConvertType={props.searchOptions.kanaConvertType}
            onChange={handleKanaConvertTypeChange}
        />
    );
};

export default SearchOptionsInput;
