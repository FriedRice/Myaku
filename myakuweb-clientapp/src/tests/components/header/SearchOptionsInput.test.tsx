/**
 * Tests for the SearchOptionsInput component.
 * @module tests/components/header/SearchOptionsInput.test
 */

import { KanaConvertType } from 'ts/types/types';
import KanaConvertTypeInput from 'ts/components/header/KanaConvertTypeInput';
import React from 'react';
import SearchOptionsInput from 'ts/components/header/SearchOptionsInput';
import { expectComponent } from 'tests/testUtils';

import {
    ShallowWrapper,
    shallow,
} from 'enzyme';

function callKanaConvertTypeInputOnChange(
    wrapper: ShallowWrapper, value: KanaConvertType
): void {
    const input = wrapper.find(KanaConvertTypeInput);
    expect(input).toHaveLength(1);
    input.props().onChange(value);
}


describe('<SearchOptionsInput />', function() {
    it('renders correctly', function() {
        const wrapper = shallow(
            <SearchOptionsInput
                searchOptions={{'kanaConvertType': 'hira'}}
                onChange={(): void => {}}
            />
        );
        expect(wrapper).toMatchSnapshot();
    });

    it('forwards KanaConvertType value', function() {
        const wrapper = shallow(
            <SearchOptionsInput
                searchOptions={{'kanaConvertType': 'hira'}}
                onChange={(): void => {}}
            />
        );
        expectComponent(wrapper, KanaConvertTypeInput, {
            kanaConvertType: 'hira',
            onChange: expect.any(Function),
        });

        wrapper.setProps({'searchOptions': {'kanaConvertType': 'kata'}});
        expectComponent(wrapper, KanaConvertTypeInput, {
            kanaConvertType: 'kata',
            onChange: expect.any(Function),
        });

        wrapper.setProps({'searchOptions': {'kanaConvertType': 'none'}});
        expectComponent(wrapper, KanaConvertTypeInput, {
            kanaConvertType: 'none',
            onChange: expect.any(Function),
        });

        wrapper.setProps({'searchOptions': {'kanaConvertType': 'hira'}});
        expectComponent(wrapper, KanaConvertTypeInput, {
            kanaConvertType: 'hira',
            onChange: expect.any(Function),
        });
    });

    it('calls onChange when KanaConvertType onChange called', function() {
        const mockOnChange = jest.fn();
        const wrapper = shallow(
            <SearchOptionsInput
                searchOptions={{'kanaConvertType': 'hira'}}
                onChange={mockOnChange}
            />
        );

        callKanaConvertTypeInputOnChange(wrapper, 'kata');
        expect(mockOnChange).toBeCalledTimes(1);
        expect(mockOnChange).lastCalledWith('kanaConvertType', 'kata');

        callKanaConvertTypeInputOnChange(wrapper, 'none');
        expect(mockOnChange).toBeCalledTimes(2);
        expect(mockOnChange).lastCalledWith('kanaConvertType', 'none');

        callKanaConvertTypeInputOnChange(wrapper, 'hira');
        expect(mockOnChange).toBeCalledTimes(3);
        expect(mockOnChange).lastCalledWith('kanaConvertType', 'hira');
    });
});
