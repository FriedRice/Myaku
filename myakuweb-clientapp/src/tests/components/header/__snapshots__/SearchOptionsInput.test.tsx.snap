// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchOptionsInput /> renders correctly 1`] = `
<KanaConvertTypeInput
  kanaConvertType="hira"
  onChange={[Function]}
/>
`;
